tests/integration/test_storage_nats/test_nats_jet_stream.py

```diff
@@ -0,0 +1,1766 @@
+        result = instance.query("SELECT count(DISTINCT key) FROM test.view")
+        time.sleep(1)
+        if int(result) == messages_num:
+            break
+
+    assert int(result) == messages_num, "ClickHouse lost some messages: {}".format(
+        result
+    )
+
+
+def test_nats_no_connection_at_startup_1(nats_cluster):
+
+    asyncio.run(add_durable_consumer(cluster, "test_stream", "test_consumer"))
+
+    instance.query_and_get_error(
+        f"""
+        CREATE TABLE test.cs (key UInt64, value UInt64)
+            ENGINE = NATS
+            SETTINGS nats_url = 'invalid_nats_url:4444',
+                    nats_stream = 'test_stream',
+                    nats_consumer_name = 'test_consumer',
+                    nats_subjects = 'test_subject',
+                    nats_format = 'JSONEachRow',
+                    nats_num_consumers = '5',
+                    nats_row_delimiter = '\\n';
+        """
+    )
+    instance.query_and_get_error(
+        """
+        SHOW TABLE test.cs;
+        """
+    )
+
+
+def test_nats_no_connection_at_startup_2(nats_cluster):
+
+    asyncio.run(add_durable_consumer(cluster, "test_stream", "test_consumer"))
+
+    instance.query(
+        f"""
+        CREATE TABLE test.cs (key UInt64, value UInt64)
+            ENGINE = NATS
+            SETTINGS nats_url = 'nats1:4444',
+                     nats_stream = 'test_stream',
+                     nats_consumer_name = 'test_consumer',
+                     nats_subjects = 'test_subject',
+                     nats_format = 'JSONEachRow',
+                     nats_num_consumers = '5',
+                     nats_row_delimiter = '\\n';
+        CREATE TABLE test.view (key UInt64, value UInt64)
+            ENGINE = MergeTree
+            ORDER BY key;
+        CREATE MATERIALIZED VIEW test.consumer TO test.view AS
+            SELECT * FROM test.cs;
+        """
+    )
+
+    instance.query(
+        """
+        DETACH TABLE test.cs;
+        DROP VIEW test.consumer;
+        """
+    )
+    with nats_cluster.pause_container("nats1"):
+        nats_helpers.wait_nats_paused(nats_cluster.nats_port, nats_cluster.nats_ssl_context)
+        instance.query("ATTACH TABLE test.cs")
+
+    asyncio.run(add_durable_consumer(cluster, "test_stream", "test_consumer"))
+
+    instance.query(
+        """
+        CREATE MATERIALIZED VIEW test.consumer TO test.view AS
+            SELECT * FROM test.cs;
+        """
+    )
+
+    messages_num = 1000
+    messages = []
+    for i in range(messages_num):
+        messages.append(json.dumps({"key": i, "value": i}))
+    asyncio.run(produce_messages(nats_cluster, "test_stream", "test_subject", messages))
+
+    for _ in range(20):
+        result = instance.query("SELECT count() FROM test.view")
+        if int(result) == messages_num:
+            break
+        time.sleep(1)
+
+    assert int(result) == messages_num, "ClickHouse lost some messages: {}".format(
+        result
+    )
```

Comment:
This is similar to the other while loops, but with a definite number of retries, I think it is better approach.